juju/juju
# [JUJU-417] Set series updates charm origin platform

This is a partial update for current deployments while more in-depth work is done on upgrade-series and charmhub charms.  For this reason, charm url is not updated in this pr as it would require downing the new charm.

To ensure that `juju refresh` finds the correct binaries after `juju upgrade-series` is run, update `set-series` to up change the application's charm origin platform series as well.  For subordinates too.  This is very important for charmhub charms, not charm store nor local charms.  In that case there should be no impact to end results.

There is a known issue where juju refresh will not give expected results if the series of a charm is forced.

A warning note has been added to `juju set-series` to refresh the charm and any subordinates before adding a new unit.

## QA steps

```console
$ juju deploy ubuntu --series bionic
$ juju set-series ubuntu focal
WARNING To ensure the correct charm binaries are installed when add-unit is next called, please first run `juju refresh` for this application and any related subordinates.
$ juju-db.bash
juju:PRIMARY> db.applications.find({"name":"ubuntu"}).pretty()
{
	...
        "name" : "ubuntu",
	...
	"series" : "focal",
	...
	"charmurl" : "ch:amd64/bionic/ubuntu-19",
	....
	"charm-origin" : {
		...
		"platform" : {
			"architecture" : "amd64",
			"os" : "ubuntu",
			"series" : "focal"
		}
	},
	...
}
$ juju refresh ubuntu
Added charm-hub charm "ubuntu", revision 19 in channel stable, to the model
$ juju-db.bash
db.applications.find({"name":"ubuntu"},{"charmurl" :1}).pretty()
{
	"_id" : "f160b2f5-a09c-4f44-8a51-983dbb36366f:ubuntu",
	"charmurl" : "ch:amd64/focal/ubuntu-19"
}
$ juju status
Model    Controller  Cloud/Region         Version   SLA          Timestamp
default  tuesday     localhost/localhost  2.9.24.1  unsupported  15:05:21Z

App     Version  Status  Scale  Charm   Store     Channel  Rev  OS      Message
ubuntu           active      2  ubuntu  charmhub  stable    19  ubuntu

Unit       Workload  Agent  Machine  Public address  Ports  Message
ubuntu/0*  active    idle   0        x.x.x.x
ubuntu/1   active    idle   1        x.x.x.x

Machine  State    DNS            Inst id        Series  AZ  Message
0        started  10.74.147.107  juju-36366f-0  bionic      Running
1        started  10.74.147.36   juju-36366f-1  focal       Running

Test the following scenarios with both a 2.8 juju and a 2.9 juju.
1. add a new model
2. deploy ubuntu rev 18 and with series = bionic, relate
3. deploy ntp with series = bionic, relate to ubuntu
4. set-series of ubuntu to focal
5. migrate the model to a controller running the test code.
6. refresh ubuntu 
7. add a new ubuntu unit - verify it is using the newer focal charm
8. repeat steps 1-4 on older controller
9. upgrade-controller, upgrade-juju
10. repeat steps 6 and 7.
```
